domain/schema/secret.go

```diff
@@ -0,0 +1,282 @@
+    );
+
+CREATE UNIQUE INDEX idx_secret_application_owned_label ON secret_application_owned (label,application_uuid);
+
+-- CREATE TRIGGER secret_application_owned_label_unique
+--     BEFORE INSERT ON secret_application_owned
+--     FOR EACH ROW
+--     WHEN (NEW.label IS NOT NULL)
+--     BEGIN
+--         SELECT RAISE(ABORT, 'label must be unique.')
+--         WHERE EXISTS (
+--             SELECT 1 FROM secret_unit_owned
+--             WHERE label = NEW.label AND unit_uuid IN (unit_uuid1, unit_uuid2, ...)
+--         );
+--     END;
+
+-- TODO: same here, we need to ensure the label is unique for the application
+-- and all its units ! ! !
+CREATE TABLE secret_unit_owned (
+        secret_metadata_uuid TEXT NOT NULL,
+        unit_uuid TEXT NOT NULL,
+        label TEXT,
+        CONSTRAINT pk_secret_unit_owned_secret_metadata_uuid_unit_uuid
+            PRIMARY KEY (secret_metadata_uuid, unit_uuid),
+        CONSTRAINT fk_secret_unit_owned_secret_metadata_uuid
+            FOREIGN KEY (secret_metadata_uuid)
+            REFERENCES secret_metadata (uuid),
+        CONSTRAINT fk_secret_unit_owned_unit_uuid
+            FOREIGN KEY (unit_uuid)
+            REFERENCES unit (uuid)
+    );
+
+CREATE UNIQUE INDEX idx_secret_unit_owned_label ON secret_unit_owned (label,unit_uuid);
+
+CREATE TABLE
+    secret_model_owned (
+        secret_metadata_uuid TEXT PRIMARY KEY,
+        label TEXT,
+        CONSTRAINT fk_secret_model_owned_secret_metadata_uuid
+            FOREIGN KEY (secret_metadata_uuid)
+            REFERENCES secret_metadata (uuid)
+    );
+
+CREATE UNIQUE INDEX idx_secret_model_owned_label ON secret_model_owned (label);
+
+-- It's ok to duplicate this table for each model database.
+CREATE TABLE
+    rotate_policy (
+        id INT PRIMARY KEY,
+        policy TEXT NOT NULL,
+        CONSTRAINT uniq_rotate_policy_policy
+            UNIQUE(policy)
+    );
+
+INSERT INTO rotate_policy VALUES
+    (0, 'never'),
+    (1, 'hourly'),
+    (2, 'daily'),
+    (3, 'weekly'),
+    (4, 'monthly'),
+    (5, 'quarterly'),
+    (6, 'yearly');
+
+CREATE TABLE
+    secret_metadata (
+        uuid TEXT PRIMARY KEY,
+        version INT,
+        description TEXT,
+        latest_revision INT NOT NULL,
+        latest_expire_time DATETIME,
+        rotate_policy TEXT,
+        auto_prune BOOLEAN NOT NULL DEFAULT (FALSE),
+        create_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        update_time DATETIME NOT NULL DEFAULT (STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'utc')),
+        CONSTRAINT fk_secret_metadata_rotate_policy
+            FOREIGN KEY (rotate_policy)
+            REFERENCES rotate_policy (policy)
+    );
+
+CREATE TABLE
+    secret_value_ref (
+        uuid TEXT PRIMARY KEY,
+        backend_id TEXT NOT NULL,
+        revision_id TEXT NOT NULL
+    );
+
+CREATE TABLE
+    secret_revision (
+        metadata_uuid TEXT NOT NULL,
+        revision INT NOT NULL,
+        data BLOB,
```

Comment:
We should avoid blobs where possible. We know the content of a secret is always key value. So we can model that.